### PR TITLE
refactor: clone default http transport for registry client

### LINKED ISSUE #839: refactor `net.Dialer` so registry client can use transport cloned from http.DefaultTransport
Need to refactor below struct to accept `DialContext` instead of `*net.Dialer` to make it further optimal.
https://github.com/oras-project/oras/blob/a98931bc68a0e7a5de6c51df8e5aa09aadad3057/internal/net/net.go#L25-L28 

_Originally posted by @TerryHowe in https://github.com/oras-project/oras/pull/838#discussion_r1119193586_

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR removes hard-coded transport declaration in registry client.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #839 

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: internal/net/net.go [hunk 1]
```diff
@@ -21,10 +21,12 @@ import (
 	"net"
 )
 
+type DialFunc func(ctx context.Context, network, addr string) (net.Conn, error)
```
Comment (shizhMSFT) [inline internal/net/net.go:24]: missing doc

**Review by TerryHowe — APPROVED** [review]
LGTM

File: cmd/oras/internal/option/remote.go [hunk 2]
```diff
@@ -193,28 +188,12 @@ func (opts *Remote) authClient(registry string, debug bool) (client *auth.Client
...
-		}),
-		ForceAttemptHTTP2:     true,
-		MaxIdleConns:          100,
-		IdleConnTimeout:       90 * time.Second,
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
-		TLSClientConfig:       config,
+	} else {
+		baseTransport.DialContext = dialContext
 	}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/remote.go:197]: nit: code style: unnecessary `else`.

File: internal/net/net.go [hunk 3]
```diff
@@ -21,10 +21,13 @@ import (
 	"net"
 )
 
+// Function type of DialContext.
+type DialFunc func(ctx context.Context, network, addr string) (net.Conn, error)
```
Comment (shizhMSFT) [inline internal/net/net.go:25]: Documentation should start wtih `// DialFunc`.

**Review by TerryHowe — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**